request_handler.py

```diff
@@ -0,0 +1,186 @@
+
+    def _get_oauth_token(self, code):
+        state = self._rsh.load_state()
+
+        client_id = state['client_id']
+        client_secret = state['client_secret']
+        redirect_uri = state['redirect_url']
+        request_url = state['request_url']
+        proxy = state['proxy']
+
+        body = {
+            'grant_type': 'authorization_code',
+            'redirect_uri': redirect_uri,
+            'client_id': client_id,
+            'code': code,
+            'client_secret': client_secret,
+        }
+        try:
+            # Ignoring request-sensitive-data semgrep check as it is a false positive
+            r = requests.post(f'{request_url}/token', data=body, proxies=proxy, timeout=60)  # nosemgrep
+            r.raise_for_status()
+            resp_json = r.json()
+        except Exception as e:
+            return False, self._return_error(
+                f"Error retrieving OAuth Token: {str(e)}",
+                401
+            )
+        state['oauth_token'] = resp_json
+        self._rsh.save_state(state)
+
+        return True, None
+
+    def is_valid_asset_id(self, asset_id):
+        """ This function validates an asset id.
+        Must be an alphanumeric string of less than 128 characters.
+
+        :param asset_id: asset_id
+        :return: is_valid: Boolean True if valid, False if not.
+        """
+        return isinstance(asset_id, str) and asset_id.isalnum() and len(asset_id) <= 128
+
+    def handle_request(self):
+        try:
+            GET = self._request.GET
+            asset_id = GET.get('state')
+            if not self.is_valid_asset_id(asset_id):
+                return self._return_error("Invalid asset id provided", 401)
+            self._rsh = RequestStateHandler(asset_id)
+            error = GET.get('error')
+            if error:
+                error_msg = GET.get('error_description')
+                return self._return_error(error_msg, 401)
+
+            code = GET.get('code')
+
+            ret_val, http_object = self._get_oauth_token(code)
+
+            if ret_val is False:
+                return http_object
+
+            return HttpResponse("You can now close this page", content_type="text/plain")
+        except Exception as e:
+            return self._return_error(f"Error handling request: {str(e)}", 400)
+
+
+class RequestStateHandler:
+    def __init__(self, asset_id):
+        asset_id = str(asset_id)
+        if asset_id and asset_id.isalnum():
+            self._asset_id = asset_id
+        else:
+            raise AttributeError("RequestStateHandler got invalid asset_id")
+
+    def _encrypt_state(self, state):
+        try:
+            if 'oauth_token' in state:
+                for token_name in consts.BMCREMEDY_TOKEN_LIST:
+                    if state['oauth_token'].get(token_name):
+                        state['oauth_token'][token_name] = encryption_helper.encrypt(  # pylint: disable=E1101
```

Comment:
`try:` should be at this level

Reply:
Same below

Reply:
Will add try in specific lavel.

Reply:
I have added `try:` at the specific level.